src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -680,14 +686,35 @@ CompilationTrackerState GetFrozenCompilationState()
                     {
                         // We have no data at all. Create a frozen empty project/compilation to represent this state.
                         var compilationWithoutGeneratedDocuments = this.CreateEmptyCompilation();
+                        var frozenProjectState = projectState.RemoveAllDocuments();
+
+                        // However, we may have already parsed some of the documents in this compilation.  For example,
+                        // if we're partway through the logic in BuildInProgressStateFromNoCompilationStateAsync.  If
+                        // so, move those parsed documents over to the new project state so we can preserve as much
+                        // information as possible.
+
+                        using var _1 = ArrayBuilder<DocumentState>.GetInstance(out var documentsWithTrees);
+                        using var _2 = ArrayBuilder<SyntaxTree>.GetInstance(out var alreadyParsedTrees);
+
+                        foreach (var (documentId, documentState) in projectState.DocumentStates.States)
```

Comment:
```suggestion
                        foreach (var (documentId, documentState) in projectState.DocumentStates.GetStatesInCompilationOrder())
```

We don't do a great job ensuring the tree order matters, and I'm not sure I care much, but if we have a convenient helper for it then maybe we should at least use it. :smile:

Reply:
>  It's far more cognitive load to understand the structure than just one bigger method here.

will change!